<!DOCTYPE html>
<html lang="en">
  <head>
    <title>AEGeAn Toolkit</title>
    <meta charset="UTF-8" />
    <meta name="description" content="AEGeAn: integrated toolkit for analysis and evaluation of annotated genomes" />
    <meta name="author" content="Daniel S. Standage" />
    <link rel="stylesheet" type="text/css" href="resources/basic-profile.css" title="Basic Profile" media="all" />
  </head>

  <body>
  <div id="wrap">
    <div id="sidebar">
      <a href="index.html"><img src="resources/aegean-logo-v1.0-256.png" alt="AEGeAn logo" style="margin-bottom: 30px;" /></a>
      
      <ul>
        <li><a href="https://github.com/standage/AEGeAn" title="AEGeAn at GitHub">Source code</a><br />AEGeAn source code at GitHub</li>
        <li><a href="https://github.com/standage/AEGeAn/releases" title="Stable releases of the AEGeAn toolkit">Releases</a><br />Stable toolkit releases</li>
        <li><a href="javascript:alert('Coming soon!')" title="Documentation for AEGeAn">Documentation</a><br />Read the docs</li>
        <li><a href="http://standage.public.iastate.edu/" title="Daniel Standage's CV">Author</a><br />Daniel Standage</li>
      </ul>
    </div>
  
    <div id="content">
      <h2>Integrated genome analysis toolkit</h2>
      <p>
        The AEGeAn Toolkit is designed for the <strong>A</strong>nalysis and <strong>E</strong>valuation of <strong>Ge</strong>nome <strong>An</strong>notations. The toolkit includes a variety of analysis programs as well as a C library whose API provides access to AEGeAn's core functions and data structures. AEGeAn is free to use and is released under an unrestrictive open source <a href="https://github.com/standage/AEGeAn/blob/master/docs/LICENSE" title="AEGeAn license">license</a>.
      </p>

      <h3>Tools</h3>
      <p>
        The AEGeAn Toolkit started as several distinct but related efforts to develop tools for managing and analyzing whole-genome gene structure annotations. AEGeAn brings these efforts together into one integrated package.
      </p>
      <ul style="margin-left: 50px;">
        <li><strong>CanonGFF3</strong> is a preprocessing tool for GFF3 files&mdash;it singles out features related to canonical protein-coding genes and validates the gene models annotated by these features.</li>
        <li><strong>LocusPocus</strong> computes gene loci based on overlapping gene models from one or more sources of annotation.</li>
        <li><strong>ParsEval</strong> compares gene structure annotations from two alternative sources of annotation for the same sequence.</li>
        <li><strong>VAnG</strong> validates a GFF3 file against a schema (in the early stages of development).</li>
      </ul>
      <p>Several additional related tools are under development, and will be integrated and documented when they are a bit more stable.</p>

      <h3>Documentation</h3>
      <p>Comprehensive documentation for the AEGeAn Toolkit is coming soon! In the mean time, installation instructions are available with the source code distribution, each program provides a descriptive usage statement, and the C API is documented in the header files located in the <code>inc/core</code> directory.</p>

      <h3>Publications</h3>
      <p><strong>Daniel S. Standage</strong> and Volker P. Brendel (2012) ParsEval: parallel comparison and analysis of gene structure annotations. <em>BMC Bioinformatics</em>, <strong>13</strong>:187, <a href="http://dx.doi.org/10.1186/1471-2105-13-187" target="_blank">doi:10.1186/1471-2105-13-187</a>.</p>

      <h3>Support</h3>
      <p>If you have any questions or run into any issues running the software, please use <a href="https://github.com/standage/AEGeAn/issues" title="Issue tracker for AEGeAn on GitHub">AEGeAn's issue tracker on GitHub</a>. I receive an email notification with each issue submitted and will do my best to respond quickly.</p>

      <p class="credits">&copy; 2010-2013 Daniel S. Standage | Template design by <a href="http://andreasviklund.com/">Andreas Viklund</a></p>
    </div>
  </div>
  </body>
</html>
