<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <title>ParsEval Summary</title>
    <link rel="stylesheet" type="text/css" href="parseval.css" />
    <script type="text/javascript" language="javascript" src="vendor/jquery.js"></script>
    <script type="text/javascript" language="javascript" src="vendor/jquery.dataTables.js"></script>
    <script type="text/javascript">
      $(document).ready(function() {
        $('#seqlist').dataTable( {
          "sScrollY": "400px",
          "bPaginate": false,
          "bScrollCollapse": true,
          "bSort": false,
          "bFilter": false,
          "bInfo": false
        });
      } );
    </script>
  </head>
  <body>
    <div id="content">
      <h1>ParsEval Summary</h1>
      <pre class="command">
Started:                03 Sep 2014, 10:35AM
Reference annotations:  OGS
Prediction annotations: NCBI
Executing command:      bin/parseval --datashare=data/share/ --outformat=html --png --outfile=pe.temp --overwrite --refrlabel=OGS --predlabel=NCBI data/gff3/amel-ogs-g716.gff3 data/gff3/amel-ncbi-g716.gff3 </pre>

      <h2>Sequences compared</h2>
      <p class="indent">Click on a sequence ID below to see comparison results for individual loci.</p>
      <table id="seqlist" class="indent">
        <thead>
          <tr>
            <th>Sequence</th>
            <th>Refr genes</th>
            <th>Pred genes</th>
            <th>Loci</th>
          </tr>
        </thead>
        <tbody>
        <tr><td><a href="Group7.16/index.html">Group7.16</a></td><td>10</td><td>8</td><td>8</td></tr>
        </tbody>

      </table>

      <h2>Gene loci <span class="tooltip">[?]<span class="tooltip_text">If a gene annotation overlaps with another gene annotation, those annotations are associated with the same gene locus. See <a target="_blank" href="http://aegean.readthedocs.org/en/refactor/loci.html">this page</a> for a formal definition of a locus annotation.</span></span></h2>
      <table class="table_normal">
        <tr><td>shared</td><td>7</td></tr>
        <tr><td>unique to reference</td><td>1</td></tr>
        <tr><td>unique to prediction</td><td>0</td></tr>
        <tr><th class="right-align">Total</th><td>8</td></tr>
      </table>

      <h2>Reference annotations</h2>
      <table class="table_normal">
        <tr><td>genes</td><td>10</td></tr>
        <tr class="cell_small"><td class="cell_indent">average per locus</td><td>1.250</td></tr>
        <tr><td>transcripts</td><td>10</td></tr>
        <tr class="cell_small"><td class="cell_indent">average per locus</td><td>1.250</td></tr>
        <tr class="cell_small"><td class="cell_indent">average per gene</td><td>1.000</td></tr>
      </table>

      <h2>Prediction annotations</h2>
      <table class="table_normal">
        <tr><td>genes</td><td>8</td></tr>
        <tr class="cell_small"><td class="cell_indent">average per locus</td><td>1.000</td></tr>
        <tr><td>transcripts</td><td>10</td></tr>
        <tr class="cell_small"><td class="cell_indent">average per locus</td><td>1.250</td></tr>
        <tr class="cell_small"><td class="cell_indent">average per gene</td><td>1.250</td></tr>
      </table>

      <h2>Comparisons</h2>
      <table class="comparisons">
        <tr><th>Total comparisons</th><th>8</th></tr>
        <tr><td>perfect matches <span class="tooltip"><span class="small_tooltip">[?]</span><span class="tooltip_text">Prediction transcripts (exons, coding sequences, andUTRs) line up perfectly with reference transcripts.</span></span></td><td>0 (0.0%)</tr>
        <tr><td>perfect matches with mislabeled UTRs <span class="tooltip"><span class="small_tooltip">[?]</span><span class="tooltip_text">5'/3' orientation of UTRs is reversed between reference and prediction, but a perfect match in all other aspects.</span></span></td><td>0 (0.0%)</tr>
        <tr><td>CDS structure matches <span class="tooltip"><span class="small_tooltip">[?]</span><span class="tooltip_text">Not a perfect match, but prediction coding sequence(s) line up perfectly with reference coding sequence(s).</span></span></td><td>4 (50.0%)</tr>
        <tr class="cell_small"><td class="cell_indent">average length</td><td>8691.00 bp</td></tr>
        <tr class="cell_small"><td class="cell_indent">average # refr exons</td><td>5.75</td></tr>
        <tr class="cell_small"><td class="cell_indent">average # pred exons</td><td>6.00</td></tr>
        <tr class="cell_small"><td class="cell_indent">average refr CDS length</td><td>455.25 aa</td></tr>
        <tr class="cell_small"><td class="cell_indent">average pred CDS length</td><td>455.25 aa</td></tr>
        <tr><td>exon structure matches <span class="tooltip"><span class="small_tooltip">[?]</span><span class="tooltip_text">Not a perfect match or CDS match, but prediction exon structure is identical to reference exon structure.</span></span></td><td>0 (0.0%)</tr>
        <tr><td>UTR structure matches <span class="tooltip"><span class="small_tooltip">[?]</span><span class="tooltip_text">Not a perfect match, CDS match, or exon structure match, but prediction UTRs line up perfectly with reference UTRs.</span></span></td><td>0 (0.0%)</tr>
        <tr><td>non-matches <span class="tooltip"><span class="small_tooltip">[?]</span><span class="tooltip_text">Differences in CDS, exon, and UTR structure.</span></span></td><td>4 (50.0%)</tr>
        <tr class="cell_small"><td class="cell_indent">average length</td><td>13661.75 bp</td></tr>
        <tr class="cell_small"><td class="cell_indent">average # refr exons</td><td>6.50</td></tr>
        <tr class="cell_small"><td class="cell_indent">average # pred exons</td><td>6.25</td></tr>
        <tr class="cell_small"><td class="cell_indent">average refr CDS length</td><td>398.25 aa</td></tr>
        <tr class="cell_small"><td class="cell_indent">average pred CDS length</td><td>391.50 aa</td></tr>
      </table>

      <h3>CDS structure comparison</h3>
      <table class="table_normal table_extra_indent">
        <tr><td>reference CDS segments</td><td>49</td></tr>
        <tr class="cell_small"><td class="cell_indent">match prediction</td><td>44 (89.8%)</td></tr>
        <tr class="cell_small"><td class="cell_indent">don't match prediction</td><td>5 (10.2%)</td></tr>
        <tr><td>prediction CDS segments</td><td>48</td></tr>
        <tr class="cell_small"><td class="cell_indent">match prediction</td><td>44 (91.7%)</td></tr>
        <tr class="cell_small"><td class="cell_indent">don't match prediction</td><td>4 (8.3%)</td></tr>
        <tr><td>Sensitivity</td><td>0.898</td></tr>
        <tr><td>Specificity</td><td>0.917</td></tr>
        <tr><td>F1 score</td><td>0.907</td></tr>
        <tr><td>Annotation edit distance</td><td>0.093</td></tr>
      </table>

      <h3>Exon structure comparison</h3>
      <table class="table_normal table_extra_indent">
        <tr><td>reference exons</td><td>49</td></tr>
        <tr class="cell_small"><td class="cell_indent">match prediction</td><td>33 (67.3%)</td></tr>
        <tr class="cell_small"><td class="cell_indent">don't match prediction</td><td>16 (32.7%)</td></tr>
        <tr><td>prediction exons</td><td>49</td></tr>
        <tr class="cell_small"><td class="cell_indent">match prediction</td><td>33 (67.3%)</td></tr>
        <tr class="cell_small"><td class="cell_indent">don't match prediction</td><td>16 (32.7%)</td></tr>
        <tr><td>Sensitivity</td><td>0.673</td></tr>
        <tr><td>Specificity</td><td>0.673</td></tr>
        <tr><td>F1 score</td><td>0.673</td></tr>
        <tr><td>Annotation edit distance</td><td>0.327</td></tr>
      </table>

      <h3>UTR structure comparison</h3>
      <table class="table_normal table_extra_indent">
        <tr><td>reference UTR segments</td><td>10</td></tr>
        <tr class="cell_small"><td class="cell_indent">match prediction</td><td>0 (0.0%)</td></tr>
        <tr class="cell_small"><td class="cell_indent">don't match prediction</td><td>10 (100.0%)</td></tr>
        <tr><td>prediction UTR segments</td><td>16</td></tr>
        <tr class="cell_small"><td class="cell_indent">match prediction</td><td>0 (0.0%)</td></tr>
        <tr class="cell_small"><td class="cell_indent">don't match prediction</td><td>16 (100.0%)</td></tr>
        <tr><td>Sensitivity</td><td>0.000</td></tr>
        <tr><td>Specificity</td><td>0.000</td></tr>
        <tr><td>F1 score</td><td>nan</td></tr>
        <tr><td>Annotation edit distance</td><td>1.000</td></tr>
      </table>

      <h3>Nucleotide-level comparison</h3>
      <table class="table_wide table_extra_indent">
        <tr><th>&nbsp;</th><th>CDS</th><th>UTRs</th><th>Overall</th></tr>
        <tr><th class="left-align">matching coefficient</th><td>0.995</td><td>0.964</td><td>0.802</td></tr>
        <tr><th class="left-align">correlation coefficient</th><td>0.978</td><td>0.430</td><td>--</td></tr>
        <tr><th class="left-align">sensitivity</th><td>0.976</td><td>0.371</td><td>--</td></tr>
        <tr><th class="left-align">specificity</th><td>0.984</td><td>0.540</td><td>--</td></tr>
        <tr><th class="left-align">F1 score</th><td>0.980</td><td>0.440</td><td>--</td></tr>
        <tr><th class="left-align">annotation edit distance</th><td>0.020</td><td>0.544</td><td>--</td></tr>
      </table>

      <p class="footer">
        Generated by <a href="https://github.com/standage/AEGeAn/tree/935b76f17bf2586a9ba6a4cb17759fd0caae5829">AEGeAn version 935b76f17b</a>.<br />
        Copyright Â© 2014 <a href="http://aegean.readthedocs.org/en/latest/contrib.html">AEGeAn authors</a>.<br />
        See <a href="LICENSE">LICENSE</a> for details.      </p>
